### PR TITLE
Language support

### PR DESCRIPTION
**Pull request by scambra**
I have added support to look for movies in different language than english. It's on languages branch, I didn't want to push to master until it's completed and reviewed.

It adds another page on settings, languages, to add a language code (language-COUNTRY, ISO 639-1 and ISO 3166-1), as used by TheMovieDB. Each language code has language names, which are names for the language to require them in title, to consider movie is in requested language.

Movies can be edited to change language, then it queries themoviedb to look for titles in that language (it also changes plot with translated plot in language if available), and allows user to pick one title (sometimes multiple titles are provided in alternative titles and translations). English title is saved in new column, english_title.

When searching for torrents, 2 queries are sent, with translated title and english title. Language's names are removed from ignore words list, and one name is required to be in torrent's name to compare torrent's title with english_title.

If user request movie's original language, then english title is not saved, so movie is looked only with original title, and language's names are not required to be in the title, although they are removed from ignore words list.

@barbequesauce @ellnic thoughts?

### CODE REVIEW

**Comment by barbequesauce** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 20
           

Complexity increasing per file
==============================
- static/js/settings/languages.js  4
- core/library.py  3
         

Clones added
============
- core/ajax.py  4
- core/library.py  1
- static/js/library/status.js  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/saucebarbeque/Watcher3/pullRequest?prid=6225754&bid=20290178)

File: static/css/settings/languages.css [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#languages tr .code{
+    width: 30%;
+    display: inline-block;
+    text-align: center;
+}
+#languages tr .names{
```
Comment (barbequesauce) [inline static/css/settings/languages.css:6]: Codacy found an issue: [Expected single space before "{" (block-opening-brace-space-before)](https://app.codacy.com/manual/saucebarbeque/Watcher3/pullRequest?prid=6225754)

File: static/css/settings/languages.css [hunk 2]
```diff
@@ -0,0 +1,10 @@
+#languages tr .code{
+    width: 30%;
```
Comment (barbequesauce) [inline static/css/settings/languages.css:2]: Codacy found an issue: [Expected indentation of 2 spaces (indentation)](https://app.codacy.com/manual/saucebarbeque/Watcher3/pullRequest?prid=6225754)

File: core/ajax.py [hunk 3]
```diff
@@ -861,6 +881,37 @@ def update_metadata(self, imdbid, tmdbid=None):
         else:
             return r
 
+    @cherrypy.expose
+    @cherrypy.tools.json_out()
+    def movie_metadata(self, imdbid, tmdbid=None, language=None):
```
Comment (barbequesauce) [inline core/ajax.py:886]: Codacy found an issue: [Method could be a function](https://app.codacy.com/manual/saucebarbeque/Watcher3/pullRequest?prid=6225754)

File: static/js/settings/languages.js [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+                        $tr.remove();
+                     });
+}
+
+function _get_settings(){
+    var langs = {};
+    var blanks = false;
+
+// Categories
+    var required_fields = ["code", "names"];
```
Comment (barbequesauce) [inline static/js/settings/languages.js:27]: Codacy found an issue: ['required_fields' is defined but never used.](https://app.codacy.com/manual/saucebarbeque/Watcher3/pullRequest?prid=6225754)

File: core/app.py [hunk 5]
```diff
@@ -80,6 +80,7 @@ def defaults(self):
     search_template = Template(filename='templates/settings/search.html', module_directory=core.MAKO_CACHE)
     quality_template = Template(filename='templates/settings/quality.html', module_directory=core.MAKO_CACHE)
     categories_template = Template(filename='templates/settings/categories.html', module_directory=core.MAKO_CACHE)
+    languages_template = Template(filename='templates/settings/languages.html', module_directory=core.MAKO_CACHE)
```
Comment (barbequesauce) [inline core/app.py:83]: Codacy found an issue: [Mako templates allow HTML/JS rendering by default and are inherently open to XSS attacks. Ensure variables in all templates are properly sanitized via the 'n', 'h' or 'x' flags (depending on context). For example, to HTML escape the variable 'data' do ${ data |h }.](https://app.codacy.com/manual/saucebarbeque/Watcher3/pullRequest?prid=6225754)

File: static/css/settings/languages.css [hunk 6]
```diff
@@ -0,0 +1,10 @@
...
+    display: inline-block;
+    text-align: center;
```
Comment (barbequesauce) [inline static/css/settings/languages.css:4]: Codacy found an issue: [Expected indentation of 2 spaces (indentation)](https://app.codacy.com/manual/saucebarbeque/Watcher3/pullRequest?prid=6225754)

File: static/js/settings/languages.js [hunk 7]
```diff
@@ -0,0 +1,52 @@
+/*global each, category_template, $categories_form */
```
Comment (barbequesauce) [inline static/js/settings/languages.js:1]: Codacy found an issue: ['category_template' is defined but never used.](https://app.codacy.com/manual/saucebarbeque/Watcher3/pullRequest?prid=6225754)

File: static/js/settings/languages.js [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+function delete_language(event, button){
+    event.preventDefault();
+    var $tr = $(button).closest('tr');
+    $tr.fadeOut(500, function(){
+                        $tr.remove();
+                     });
+}
+
+function _get_settings(){
```
Comment (barbequesauce) [inline static/js/settings/languages.js:22]: Codacy found an issue: ['_get_settings' is defined but never used.](https://app.codacy.com/manual/saucebarbeque/Watcher3/pullRequest?prid=6225754)

File: static/js/settings/languages.js [hunk 9]
```diff
@@ -0,0 +1,52 @@
+/*global each, category_template, $categories_form */
+window.addEventListener("DOMContentLoaded", function(){
+    language_template = document.querySelector("template#language_template").innerHTML;
+    $categories_form = $("form#languages tbody");
+});
+
+function add_language(event){
```
Comment (barbequesauce) [inline static/js/settings/languages.js:7]: Codacy found an issue: ['add_language' is defined but never used.](https://app.codacy.com/manual/saucebarbeque/Watcher3/pullRequest?prid=6225754)

File: static/css/settings/languages.css [hunk 10]
```diff
@@ -0,0 +1,10 @@
+#languages tr .code{
+    width: 30%;
+    display: inline-block;
```
Comment (barbequesauce) [inline static/css/settings/languages.css:3]: Codacy found an issue: [Expected indentation of 2 spaces (indentation)](https://app.codacy.com/manual/saucebarbeque/Watcher3/pullRequest?prid=6225754)

File: static/js/settings/languages.js [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+            return;
+        }
+        var names = element.querySelector("input[data-id=\"names\"]").value;
+        if(!names){
+            element.querySelector("input[data-id=\"names\"]").classList.add("border-danger");
+            blanks = true;
+            return;
+        }
+
+        langs[code] = names;
```
Comment (barbequesauce) [inline static/js/settings/languages.js:44]: Codacy found an issue: [Generic Object Injection Sink](https://app.codacy.com/manual/saucebarbeque/Watcher3/pullRequest?prid=6225754)

File: static/js/settings/languages.js [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+function add_language(event){
+    event.preventDefault();
+    var $new_language = $(language_template);
+    $categories_form.append($new_language);
+    $new_language.find("[data-toggle=\"tooltip\"]").tooltip();
+}
+
+function delete_language(event, button){
+    event.preventDefault();
+    var $tr = $(button).closest('tr');
```
Comment (barbequesauce) [inline static/js/settings/languages.js:16]: Codacy found an issue: [Strings must use doublequote.](https://app.codacy.com/manual/saucebarbeque/Watcher3/pullRequest?prid=6225754)

File: static/css/settings/languages.css (hunk 1 above)
Comment (barbequesauce) [inline static/css/settings/languages.css:6]: Codacy found an issue: [Expected empty line before rule (rule-empty-line-before)](https://app.codacy.com/manual/saucebarbeque/Watcher3/pullRequest?prid=6225754)

File: static/css/settings/languages.css [hunk 13]
```diff
@@ -0,0 +1,10 @@
+#languages tr .code{
```
Comment (barbequesauce) [inline static/css/settings/languages.css:1]: Codacy found an issue: [Expected single space before "{" (block-opening-brace-space-before)](https://app.codacy.com/manual/saucebarbeque/Watcher3/pullRequest?prid=6225754)

File: static/js/library/status.js [hunk 14]
```diff
@@ -602,6 +613,47 @@ function update_metadata(event, elem, imdbid, tmdbid){
     });
 }
 
+function change_language(event, elem, imdbid, tmdbid){
```
Comment (barbequesauce) [inline static/js/library/status.js:616]: Codacy found an issue: ['change_language' is defined but never used.](https://app.codacy.com/manual/saucebarbeque/Watcher3/pullRequest?prid=6225754)

File: static/js/settings/languages.js [hunk 15]
```diff
@@ -0,0 +1,52 @@
+/*global each, category_template, $categories_form */
+window.addEventListener("DOMContentLoaded", function(){
+    language_template = document.querySelector("template#language_template").innerHTML;
```
Comment (barbequesauce) [inline static/js/settings/languages.js:3]: Codacy found an issue: ['language_template' is not defined.](https://app.codacy.com/manual/saucebarbeque/Watcher3/pullRequest?prid=6225754)

File: static/js/settings/languages.js [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+    $categories_form = $("form#languages tbody");
+});
+
+function add_language(event){
+    event.preventDefault();
+    var $new_language = $(language_template);
```
Comment (barbequesauce) [inline static/js/settings/languages.js:9]: Codacy found an issue: ['language_template' is not defined.](https://app.codacy.com/manual/saucebarbeque/Watcher3/pullRequest?prid=6225754)

File: static/js/settings/languages.js [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+    $categories_form.append($new_language);
+    $new_language.find("[data-toggle=\"tooltip\"]").tooltip();
+}
+
+function delete_language(event, button){
```
Comment (barbequesauce) [inline static/js/settings/languages.js:14]: Codacy found an issue: ['delete_language' is defined but never used.](https://app.codacy.com/manual/saucebarbeque/Watcher3/pullRequest?prid=6225754)

File: static/css/settings/languages.css [hunk 18]
```diff
@@ -0,0 +1,10 @@
...
+    width: 30%;
+    display: inline-block;
+    text-align: center;
+}
+#languages tr .names{
+    width: 69%;
```
Comment (barbequesauce) [inline static/css/settings/languages.css:7]: Codacy found an issue: [Expected indentation of 2 spaces (indentation)](https://app.codacy.com/manual/saucebarbeque/Watcher3/pullRequest?prid=6225754)

**Comment by ellnic** [comment]
Nice job, Sergio :-) This is a much needed feature. If it’s working as expected, I don’t see any reason why it can’t be merged to master straight away.

**Comment by scambra** [comment]
I had to change arguments for search method, it needs more movie's info such as english title and language, so I changed to get whole movie dict, and changed every method calling this. That's why I didn't push to master directly, it could break something, although I didn't see any issue in last few days. Maybe someone else could test it, no need to use languages, but ensure it still works as expected for english.

Here is one screenshot of UI changes in movie window
image

I haven't added language to add movie window, I can add it later

**Comment by scambra** [comment]
@barbequesauce @ellnic do you want to test this branch? or is it ok to merge into master?

**Comment by barbequesauce** [comment]
Yeah sorry I've been running it since Saturday - no real issues - I'm good with it, great work thank you!

**Comment by ellnic** [comment]
Sorry for the delay in replying. I'm missing a lot of notifications at the moment (massively busy at work right now). October is always pretty busy so you might have to nag me if it's urgent.

**Pull request merged**